{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dd41adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This jupyter lab allows users to create the final database needed with a single run. \n",
    "# Users can also choose to import database using the final_db.sql file provided. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13b6210d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a database named final and create the first table named pop_superpop\n",
    "import pymysql\n",
    "import mysql.connector\n",
    "from mysql.connector import Error\n",
    "from mysql.connector import errorcode\n",
    "import pandas as pd\n",
    "import zarr\n",
    "import time\n",
    "import re\n",
    "\n",
    "# Specify the database information here\n",
    "db_name = \"final\"\n",
    "db_host = \"localhost\"\n",
    "db_username = \"root\"\n",
    "db_password = \"Teammint123@\"\n",
    "\n",
    "try:\n",
    "    conn = pymysql.connect(host=db_host,\n",
    "                           port=int(3306),\n",
    "                           user=db_username,\n",
    "                           password=db_password)\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    # SQL commands for creating database and first table pop_superpop\n",
    "    sql_commands = [\n",
    "        \"CREATE DATABASE IF NOT EXISTS final;\",\n",
    "        \"USE final;\",\n",
    "        \"CREATE TABLE IF NOT EXISTS pop_superpop (sample_id VARCHAR(255) NOT NULL, population VARCHAR(255), superpopulation VARCHAR(255), PRIMARY KEY (sample_id));\"\n",
    "    ]\n",
    "\n",
    "    # Execute SQL commands\n",
    "    for command in sql_commands:\n",
    "        # Reconnect if cursor is not connected\n",
    "        if not cursor.connection:\n",
    "            cursor = conn.cursor()\n",
    "        cursor.execute(command)\n",
    "\n",
    "except pymysql.Error as e:\n",
    "    print(\"Error:\", e)\n",
    "\n",
    "finally:\n",
    "    if conn:\n",
    "        conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3659985",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populating data for table pop_superpop here.\n",
    "\n",
    "# change the path of pop_superpop.tsv file if needed\n",
    "superpop = pd.read_csv(\"pop_superpop.tsv\", sep='\\t')\n",
    "\n",
    "try:\n",
    "    # connect to the mysql server\n",
    "    connection = mysql.connector.connect(user= db_username,\n",
    "                                         password= db_password,\n",
    "                                         host= db_host,\n",
    "                                         database= db_name)\n",
    "    if connection.is_connected:\n",
    "        print(\"connection succeeded\")\n",
    "        cursor = connection.cursor()\n",
    "        # specify table name here\n",
    "        table_name = 'pop_superpop'\n",
    "        # specify table name here, for example: 'sample_id, population, superpopulation'\n",
    "        column_names = 'sample_id, population, superpopulation'\n",
    "        batch_size = 1000\n",
    "        # Use executemany to insert data in batches\n",
    "        for i in range(0, len(superpop), batch_size):\n",
    "            batch_data = superpop.iloc[i:i + batch_size].values.tolist()\n",
    "            query = f\"INSERT INTO {table_name} ({column_names}) VALUES ({', '.join(['%s'] * len(superpop.columns))})\"\n",
    "            cursor.executemany(query, batch_data)\n",
    "            connection.commit()\n",
    "\n",
    "        print('records inserted successfully!')\n",
    "\n",
    "except Error as err:\n",
    "    # check connection error\n",
    "    if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:\n",
    "        print(\"Username or password isnt correct\")\n",
    "    elif err.errno == errorcode.ER_BAD_DB_ERROR:\n",
    "        print(\"Database does not exist\")\n",
    "    else:\n",
    "        print(err)\n",
    "finally:\n",
    "    # close connection in the end\n",
    "    if connection.is_connected():\n",
    "        cursor.close()\n",
    "        connection.close()\n",
    "        print(\"MySQL connection closed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d681405",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the second table named pca_data here\n",
    "\n",
    "try:\n",
    "    conn = pymysql.connect(host=db_host,\n",
    "                           port=int(3306),\n",
    "                           user=db_username,\n",
    "                           password=db_password,\n",
    "                           db=db_name)\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    # SQL commands for creating second table pca_data\n",
    "    sql_commands = [\n",
    "        \"CREATE TABLE IF NOT EXISTS pca_data (sample_id VARCHAR(255) NOT NULL, pc1 FLOAT, pc2 FLOAT, PRIMARY KEY (sample_id), FOREIGN KEY (sample_id) REFERENCES pop_superpop (sample_id));\"\n",
    "    ]\n",
    "\n",
    "    # Execute SQL commands\n",
    "    for command in sql_commands:\n",
    "        # Reconnect if cursor is not connected\n",
    "        if not cursor.connection:\n",
    "            cursor = conn.cursor()\n",
    "        cursor.execute(command)\n",
    "\n",
    "except pymysql.Error as e:\n",
    "    print(\"Error:\", e)\n",
    "\n",
    "finally:\n",
    "    if conn:\n",
    "        conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be4eee37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populating data for table pca_data here, change the path of pca_data.csv if needed.\n",
    "clustering_data = pd.read_csv(\"pca_data.csv\", sep=',')\n",
    "\n",
    "try:\n",
    "    # connect to the mysql server\n",
    "    connection = mysql.connector.connect(user= db_username,\n",
    "                                         password= db_password,\n",
    "                                         host= db_host,\n",
    "                                         database= db_name)\n",
    "    if connection.is_connected:\n",
    "        print(\"connection succeeded\")\n",
    "        cursor = connection.cursor()\n",
    "        # specify table name here\n",
    "        table_name = 'pca_data'\n",
    "        # specify table name here, for example: 'sample_id, population, superpopulation'\n",
    "        column_names = 'sample_id, pc1, pc2'\n",
    "        batch_size = 1000\n",
    "        # Use executemany to insert data in batches\n",
    "        for i in range(0, len(clustering_data), batch_size):\n",
    "            batch_data = clustering_data.iloc[i:i + batch_size].values.tolist()\n",
    "            query = f\"INSERT INTO {table_name} ({column_names}) VALUES ({', '.join(['%s'] * len(clustering_data.columns))})\"\n",
    "            cursor.executemany(query, batch_data)\n",
    "            connection.commit()\n",
    "\n",
    "        print('records inserted successfully!')\n",
    "\n",
    "except Error as err:\n",
    "    # check connection error\n",
    "    if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:\n",
    "        print(\"Username or password isnt correct\")\n",
    "    elif err.errno == errorcode.ER_BAD_DB_ERROR:\n",
    "        print(\"Database does not exist\")\n",
    "    else:\n",
    "        print(err)\n",
    "finally:\n",
    "    # close connection in the end\n",
    "    if connection.is_connected():\n",
    "        cursor.close()\n",
    "        connection.close()\n",
    "        print(\"MySQL connection closed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56b308e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the third table named admixture_data here\n",
    "\n",
    "try:\n",
    "    conn = pymysql.connect(host=db_host,\n",
    "                           port=int(3306),\n",
    "                           user=db_username,\n",
    "                           password=db_password,\n",
    "                           db=db_name)\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    # SQL command for creating table admixture_data\n",
    "    sql_command = \"\"\"\n",
    "        CREATE TABLE IF NOT EXISTS admixture_data (\n",
    "            sample_id VARCHAR(255) NOT NULL,\n",
    "            pop1 FLOAT,\n",
    "            pop2 FLOAT,\n",
    "            pop3 FLOAT,\n",
    "            pop4 FLOAT,\n",
    "            pop5 FLOAT,\n",
    "            assignment VARCHAR(255),\n",
    "            PRIMARY KEY (sample_id),\n",
    "            FOREIGN KEY (sample_id) REFERENCES pop_superpop (sample_id)\n",
    "        );\n",
    "    \"\"\"\n",
    "\n",
    "    # Execute SQL command\n",
    "    cursor.execute(sql_command)\n",
    "\n",
    "    print(\"Table 'admixture_data' created successfully.\")\n",
    "\n",
    "except pymysql.Error as e:\n",
    "    print(\"Error:\", e)\n",
    "\n",
    "finally:\n",
    "    if conn:\n",
    "        conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375837e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populating table admixture_data here\n",
    "\n",
    "# Specify the path to df_sorted_q.csv here if needed \n",
    "admixture_data = pd.read_csv(\"df_sorted_q.csv\", sep=',')\n",
    "\n",
    "try:\n",
    "    # connect to the mysql server\n",
    "    connection = mysql.connector.connect(user= db_username,\n",
    "                                         password= db_password,\n",
    "                                         host= db_host,\n",
    "                                         database= db_name)\n",
    "    if connection.is_connected():\n",
    "        print(\"Connection succeeded\")\n",
    "        cursor = connection.cursor()\n",
    "        # specify table name here\n",
    "        table_name = 'admixture_data'\n",
    "        # specify table name here, for example: 'sample_id, population, superpopulation'\n",
    "        column_names = 'sample_id, pop1, pop2, pop3, pop4, pop5, assignment'\n",
    "        batch_size = 1000\n",
    "        total_records = len(admixture_data)\n",
    "        print(len(admixture_data))\n",
    "        records_inserted = 0\n",
    "        # Use executemany to insert data in batches\n",
    "        for i in range(0, total_records, batch_size):\n",
    "            print(i)\n",
    "            batch_data = admixture_data.iloc[i:i + batch_size].values.tolist()\n",
    "            query = f\"INSERT INTO {table_name} ({column_names}) VALUES ({', '.join(['%s'] * len(admixture_data.columns))})\"\n",
    "            cursor.executemany(query, batch_data)\n",
    "            connection.commit()\n",
    "            records_inserted += len(batch_data)\n",
    "            print(f\"{records_inserted}/{total_records} records inserted\")\n",
    "\n",
    "        print('Records inserted successfully!')\n",
    "\n",
    "except mysql.connector.Error as err:\n",
    "    # check connection error\n",
    "    if err.errno == mysql.connector.errorcode.ER_ACCESS_DENIED_ERROR:\n",
    "        print(\"Username or password isn't correct\")\n",
    "    elif err.errno == mysql.connector.errorcode.ER_BAD_DB_ERROR:\n",
    "        print(\"Database does not exist\")\n",
    "    else:\n",
    "        print(err)\n",
    "finally:\n",
    "    # close connection in the end\n",
    "    if connection.is_connected():\n",
    "        cursor.close()\n",
    "        connection.close()\n",
    "        print(\"MySQL connection closed.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b65ed715",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create table named snp_char here\n",
    "\n",
    "try:\n",
    "    conn = pymysql.connect(host=db_host,\n",
    "                           port=int(3306),\n",
    "                           user=db_username,\n",
    "                           password=db_password,\n",
    "                           db=db_name)\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    # SQL commands for creating table snp_char and INDEX for column position\n",
    "    sql_commands = [\n",
    "        \"SELECT * FROM admixture_data ORDER BY assignment ASC;\",\n",
    "        \"CREATE TABLE IF NOT EXISTS snp_char (snp_id VARCHAR(255) NOT NULL, position INT, ref_base CHAR(1), alt_base CHAR(1), disease_name LONGTEXT, classification LONGTEXT, PRIMARY KEY (snp_id));\",\n",
    "        \"CREATE INDEX idx_position ON snp_char (position);\"]\n",
    "\n",
    "    # Execute SQL commands\n",
    "    for command in sql_commands:\n",
    "        # Reconnect if cursor is not connected\n",
    "        if not cursor.connection:\n",
    "            cursor = conn.cursor()\n",
    "        cursor.execute(command)\n",
    "\n",
    "except pymysql.Error as e:\n",
    "    print(\"Error:\", e)\n",
    "\n",
    "finally:\n",
    "    if conn:\n",
    "        conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7075df59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populating table snp_char here\n",
    "\n",
    "# Read the zarr file here, change the path of zarr file if needed\n",
    "zarr_path = 'annotated.zarr'\n",
    "callset_h1k = zarr.open_group(zarr_path, mode='r')\n",
    "\n",
    "# Read the required columns from zarr file here\n",
    "pos = callset_h1k['1/variants/POS']\n",
    "alt = callset_h1k[\"1/variants/ALT\"]\n",
    "ref = callset_h1k['1/variants/REF']\n",
    "snp_id = callset_h1k[\"1/variants/ID\"]\n",
    "\n",
    "# Change the snp ID like \"rsXXXX:XXXXX:X:X\" to \"rsXXXX\"\n",
    "snp_id = [element.split(';')[0] if element.startswith('rs') else element for element in snp_id]\n",
    "\n",
    "data_to_insert = [\n",
    "    (SNP_ID, int(position), reference, alternate) for (SNP_ID, position, reference, alternate) in\n",
    "    zip(snp_id, pos, ref, alt)\n",
    "]\n",
    "\n",
    "try:\n",
    "    # connect to the mysql server\n",
    "    connection = mysql.connector.connect(user= db_username,\n",
    "                                         password= db_password,\n",
    "                                         host= db_host,\n",
    "                                         database= db_name)\n",
    "    if connection.is_connected:\n",
    "        print(\"connection succeeded\")\n",
    "        cursor = connection.cursor()\n",
    "        # specify table name here\n",
    "        table_name = 'snp_char'\n",
    "        # specify table name here, for example: 'sample_id, population, superpopulation'\n",
    "        column_names = 'snp_id, position, ref_base, alt_base'\n",
    "        query = f\"INSERT INTO {table_name} ({column_names}) VALUES ({', '.join(['%s'] * 4)})\"\n",
    "\n",
    "        # Use executemany to insert data in batches\n",
    "        batch_size = 1000\n",
    "        for i in range(0, len(data_to_insert), batch_size):\n",
    "            start = time.time()\n",
    "            batch = data_to_insert[i:i + batch_size]\n",
    "            cursor.executemany(query, batch)\n",
    "            connection.commit()\n",
    "            end = time.time()\n",
    "            print(f\"{i + batch_size} of lines added, time cost: {round(end-start, 2)} s\")\n",
    "\n",
    "        print('records inserted successfully!')\n",
    "\n",
    "except Error as err:\n",
    "    # check connection error\n",
    "    if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:\n",
    "        print(\"Username or password isnt correct\")\n",
    "    elif err.errno == errorcode.ER_BAD_DB_ERROR:\n",
    "        print(\"Database does not exist\")\n",
    "    else:\n",
    "        print(err)\n",
    "finally:\n",
    "    # close connection in the end\n",
    "    if connection.is_connected():\n",
    "        cursor.close()\n",
    "        connection.close()\n",
    "        print(\"MySQL connection closed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64701735",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create table named gene_names here\n",
    "\n",
    "try:\n",
    "    conn = pymysql.connect(host=db_host,\n",
    "                           port=int(3306),\n",
    "                           user=db_username,\n",
    "                           password=db_password,\n",
    "                           db=db_name)\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    # SQL command for creating table gene_names\n",
    "    sql_command = \"\"\"\n",
    "        CREATE TABLE gene_names (\n",
    "            id INT NOT NULL auto_increment,\n",
    "            gene_name VARCHAR(255) NOT NULL,\n",
    "            position INT,\n",
    "            PRIMARY KEY (id),\n",
    "            FOREIGN KEY (position) REFERENCES snp_char (position)\n",
    "        );\n",
    "    \"\"\"\n",
    "\n",
    "    # Execute SQL command\n",
    "    cursor.execute(sql_command)\n",
    "\n",
    "    print(\"Table 'gene_names' created successfully.\")\n",
    "\n",
    "except pymysql.Error as e:\n",
    "    print(\"Error:\", e)\n",
    "\n",
    "finally:\n",
    "    if conn:\n",
    "        conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "543384f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# This function creates the data list used for populating table gene_names\n",
    "def gene_data_insert():\n",
    "    \n",
    "    # Change the path if needed\n",
    "    callset_path = 'annotated.zarr'\n",
    "    csv_file_path = \"only_gene_names.csv\"\n",
    "    \n",
    "    print(\"Reading gene names from CSV file...\")\n",
    "    gene_name_df = pd.read_csv(csv_file_path, usecols=['Gene name', 'Gene start (bp)', 'Gene end (bp)'])\n",
    "\n",
    "    print(\"Reading data from Zarr file...\")\n",
    "    callset = zarr.open_group(callset_path, mode='r')\n",
    "    pos_data = callset['1/variants/POS'][:]\n",
    "\n",
    "    print(\"Mapping SNPs to genes with logging...\")\n",
    "    start_time = time.time()\n",
    "    data_to_insert = []\n",
    "\n",
    "    snp_df = pd.DataFrame(pos_data, columns=['POS'])\n",
    "\n",
    "    for index, row in gene_name_df.iterrows():\n",
    "        gene_name = row['Gene name']\n",
    "        start_bp = row['Gene start (bp)']\n",
    "        end_bp = row['Gene end (bp)']\n",
    "\n",
    "        # Match snps to genes based on snps' positions and genes' start and end base\n",
    "        matched_snps = snp_df[(snp_df['POS'] >= start_bp) & (snp_df['POS'] <= end_bp)]\n",
    "        print(f\"Gene: {gene_name}, Start: {start_bp}, End: {end_bp}, Matched SNPs: {len(matched_snps)}\")\n",
    "\n",
    "        for snp in matched_snps['POS']:\n",
    "            data_to_insert.append((gene_name, snp))\n",
    "\n",
    "    end_time = time.time()\n",
    "    print(f\"Total processing time: {end_time - start_time} seconds\")\n",
    "\n",
    "    return data_to_insert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d454dd74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populating table gene_names here:\n",
    "\n",
    "start0 = time.time()\n",
    "data_to_insert = gene_data_insert()\n",
    "end0 = time.time()\n",
    "print(f\"data_to_insert ready, time cost: {round(end0-start0, 2)} s\")\n",
    "\n",
    "try:\n",
    "    # connect to the mysql server\n",
    "    connection = mysql.connector.connect(user= db_username,\n",
    "                                         password= db_password,\n",
    "                                         host= db_host,\n",
    "                                         database= db_name)\n",
    "    if connection.is_connected:\n",
    "        print(\"connection succeeded\")\n",
    "        cursor = connection.cursor()\n",
    "        # specify table name here\n",
    "        table_name = 'gene_names'\n",
    "        # specify table name here, for example: 'sample_id, population, superpopulation'\n",
    "        column_names = 'gene_name, position'\n",
    "        query = f\"INSERT INTO {table_name} ({column_names}) VALUES ({', '.join(['%s'] * 2)})\"\n",
    "\n",
    "        # Use executemany to insert data in batches\n",
    "        batch_size = 1000\n",
    "        for i in range(0, len(data_to_insert), batch_size):\n",
    "            start = time.time()\n",
    "            batch = data_to_insert[i:i + batch_size]\n",
    "            cursor.executemany(query, batch)\n",
    "            connection.commit()\n",
    "            end = time.time()\n",
    "            print(f\"{i + batch_size} of lines added, time cost: {round(end-start, 2)} s\")\n",
    "\n",
    "        print('records inserted successfully!')\n",
    "\n",
    "except Error as err:\n",
    "    # check connection error\n",
    "    if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:\n",
    "        print(\"Username or password isnt correct\")\n",
    "    elif err.errno == errorcode.ER_BAD_DB_ERROR:\n",
    "        print(\"Database does not exist\")\n",
    "    else:\n",
    "        print(err)\n",
    "finally:\n",
    "    # close connection in the end\n",
    "    if connection.is_connected():\n",
    "        cursor.close()\n",
    "        connection.close()\n",
    "        print(\"MySQL connection closed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da274d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create table named allele_freq here\n",
    "\n",
    "try:\n",
    "    conn = pymysql.connect(host=db_host,\n",
    "                           port=int(3306),\n",
    "                           user=db_username,\n",
    "                           password=db_password,\n",
    "                           db=db_name)\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    # SQL commands for creating table allele_freq\n",
    "    sql_commands = \"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS allele_freq (\n",
    "        snp_id VARCHAR(255) NOT NULL, \n",
    "        SIB FLOAT, GBR FLOAT, FIN FLOAT, CHS FLOAT, PUR FLOAT, CDX FLOAT, \n",
    "        CLM FLOAT, IBS FLOAT, PEL FLOAT, PJL FLOAT, KHV FLOAT, ACB FLOAT, \n",
    "        GWD FLOAT, ESN FLOAT, BEB FLOAT, MSL FLOAT, STU FLOAT, ITU FLOAT, \n",
    "        CEU FLOAT, YRI FLOAT, CHB FLOAT, JPT FLOAT, LWK FLOAT, ASW FLOAT, \n",
    "        MXL FLOAT, TSI FLOAT, GIH FLOAT, \n",
    "        PRIMARY KEY (snp_id), FOREIGN KEY (snp_id) REFERENCES snp_char (snp_id)\n",
    "        );\n",
    "    \"\"\"\n",
    "    \n",
    "\n",
    "    # Execute SQL commands\n",
    "    for command in sql_commands:\n",
    "        # Reconnect if cursor is not connected\n",
    "        if not cursor.connection:\n",
    "            cursor = conn.cursor()\n",
    "        cursor.execute(command)\n",
    "\n",
    "except pymysql.Error as e:\n",
    "    print(\"Error:\", e)\n",
    "\n",
    "finally:\n",
    "    if conn:\n",
    "        conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "befd5b05-c0d6-4d33-a177-36a91928a997",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populating table allele_freq:\n",
    "\n",
    "# Specify the path to allele_freq_df.csv file if needed\n",
    "allele_freq_df = pd.read_csv(\"allele_freq_df.csv\", sep=',')\n",
    "# Fill the NA/NaN values with \"0\"\n",
    "allele_freq_df = allele_freq_df.fillna(0)\n",
    "\n",
    "try:\n",
    "    # connect to the mysql server\n",
    "    connection = mysql.connector.connect(user= db_username,\n",
    "                                         password= db_password,\n",
    "                                         host= db_host,\n",
    "                                         database= db_name)\n",
    "    if connection.is_connected:\n",
    "        print(\"connection succeeded\")\n",
    "        cursor = connection.cursor()\n",
    "        # specify table name here\n",
    "        table_name = 'allele_freq'\n",
    "        # specify table name here, for example: 'sample_id, population, superpopulation'\n",
    "        column_names = 'snp_id, SIB, GBR, FIN, CHS, PUR, CDX, CLM, IBS, PEL, PJL, KHV,'\\\n",
    "        ' ACB, GWD, ESN, BEB, MSL, STU, ITU, CEU, YRI, CHB, JPT, LWK, ASW, MXL, TSI, GIH'\n",
    "        batch_size = 1000\n",
    "        # Use executemany to insert data in batches\n",
    "        for i in range(0, len(allele_freq_df), batch_size):\n",
    "            batch_data = allele_freq_df.iloc[i:i + batch_size].values.tolist()\n",
    "            query = f\"INSERT INTO {table_name} ({column_names}) VALUES ({', '.join(['%s'] * len(allele_freq_df.columns))})\"\n",
    "            cursor.executemany(query, batch_data)\n",
    "            connection.commit()\n",
    "\n",
    "        print('records inserted successfully!')\n",
    "\n",
    "except Error as err:\n",
    "    # check connection error\n",
    "    if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:\n",
    "        print(\"Username or password isnt correct\")\n",
    "    elif err.errno == errorcode.ER_BAD_DB_ERROR:\n",
    "        print(\"Database does not exist\")\n",
    "    else:\n",
    "        print(err)\n",
    "finally:\n",
    "    # close connection in the end\n",
    "    if connection.is_connected():\n",
    "        cursor.close()\n",
    "        connection.close()\n",
    "        print(\"MySQL connection closed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "743841bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function creates the data list used for populating clinical information into table snp_char\n",
    "def data_inserting():\n",
    "    # Change the path if needed\n",
    "    clinvar_zarr_path = 'clinvar.zarr'\n",
    "    callset_cli = zarr.open_group(clinvar_zarr_path, mode='r')\n",
    "    cli_pos = callset_cli['1/variants/POS']\n",
    "    cli_alt = callset_cli['1/variants/ALT']\n",
    "    cli_ref = callset_cli['1/variants/REF']\n",
    "    disease_name = callset_cli['1/variants/CLNDN']\n",
    "    classification = callset_cli['1/variants/CLNSIG']\n",
    "    # We use the position, reference base and alternate base information of each snp\n",
    "    # from clinvar.vcf to match the snps from chr1.vcf\n",
    "    compare_data = [(x1, x2, x3) for x1, x2, x3 in zip(cli_pos, cli_ref, cli_alt)]\n",
    "    clinical_data = [[x1, x2, x3, x4, x5] for x1, x2, x3, x4, x5 in\n",
    "                     zip(disease_name, classification, cli_pos, cli_ref, cli_alt)]\n",
    "\n",
    "    zarr_path = 'annotated.zarr'\n",
    "    callset_h1k = zarr.open_group(zarr_path, mode='r')\n",
    "    snp_pos = callset_h1k['1/variants/POS']\n",
    "    snp_alt = callset_h1k['1/variants/ALT']\n",
    "    snp_ref = callset_h1k['1/variants/REF']\n",
    "    table_data = [(x1, x2, x3) for x1, x2, x3 in zip(snp_pos, snp_ref, snp_alt)]\n",
    "    table_data = set(table_data)\n",
    "\n",
    "    # Only the clinical information of matched snps will be inserted into database \n",
    "    element_pos = [index for index, element in enumerate(compare_data)\n",
    "                   if element in table_data]\n",
    "    insert_data = [clinical_data[i] for i in element_pos]\n",
    "    return insert_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f43f50a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populating column \"disease_name\" and \"classification\" in table snp_char here.\n",
    "\n",
    "insert_data = data_inserting()\n",
    "list_of_dicts = [{'name': sublist[0], 'classif': sublist[1], 'pos': int(sublist[2]), 'ref': sublist[3],\n",
    "                  'alt': sublist[4]} for sublist in insert_data]\n",
    "\n",
    "try:\n",
    "    # connect to the mysql server\n",
    "    connection = mysql.connector.connect(user=\"root\",\n",
    "                                         password=\"Teammint123@\",\n",
    "                                         host='localhost',\n",
    "                                         database='final')\n",
    "    if connection.is_connected:\n",
    "        print(\"connection succeeded\")\n",
    "        cursor = connection.cursor()\n",
    "        # specify query here\n",
    "        query = f\"UPDATE snp_char SET disease_name = %(name)s, classification = %(classif)s\" \\\n",
    "                f\"WHERE position = %(pos)s AND ref_base = %(ref)s AND alt_base = %(alt)s\"\n",
    "\n",
    "        start = time.time()\n",
    "        cursor.executemany(query, list_of_dicts)\n",
    "        connection.commit()\n",
    "        end = time.time()\n",
    "        print(f'records inserted successfully!, time cost: {round(end-start, 2)} s')\n",
    "\n",
    "except Error as err:\n",
    "    # check connection error\n",
    "    if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:\n",
    "        print(\"Username or password isnt correct\")\n",
    "    elif err.errno == errorcode.ER_BAD_DB_ERROR:\n",
    "        print(\"Database does not exist\")\n",
    "    else:\n",
    "        print(err)\n",
    "finally:\n",
    "    # close connection in the end\n",
    "    if connection.is_connected():\n",
    "        cursor.close()\n",
    "        connection.close()\n",
    "        print(\"MySQL connection closed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b592a42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This part of codes check if the snp id from annotated.zarr and allele_freq_df is matching to each other. \n",
    "import zarr\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "zarr_path = 'annotated.zarr' \n",
    "callset_h1k = zarr.open_group(zarr_path, mode='r')\n",
    "snp_id = callset_h1k[\"1/variants/ID\"]\n",
    "# change the snp ids from annotated.zarr, example: rsXXXX:XXXX:X:X to rsXXXX\n",
    "list1 = [element.split(';')[0] if element.startswith('rs') else element for element in snp_id]\n",
    "# get the snp ids from allele_freq dataframe\n",
    "allele_freq_df = pd.read_csv(\"allele_freq_df.csv\", sep=',')\n",
    "list2 = allele_freq_df.iloc[:, 0].tolist()\n",
    "# check if the snp ids from both dataframe is the same. if some of them are not, print the snp_ID which are different \n",
    "set1 = set(list1)\n",
    "filtered = [element for element in list2 if element not in set1]\n",
    "print(filtered)\n",
    "# If started fully from scrath 4 Id will not match - can change manually in the allele_freq_df.csv\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
